### PR TITLE
Add OPCode: PACKSTRUCT

### PR DESCRIPTION
**Pull request by afragapane**
#### Description of the Change
Add new OPCode: PACKSTRUCT as per https://github.com/neo-project/neo-vm/issues/51.

#### Test Plan
Added a unit test for PACKSTRUCT, mirroring the test for PACK.

#### Benefits
This will help allow typescript smart contracts compiled by neo-one to be deployed on the main neo network.
@dicarlo2

#### Applicable Issues
https://github.com/neo-project/neo-vm/issues/51

### CODE REVIEW

**Comment by erikzhang** [comment]
How do we reuse the PACK and NEWSTRUCT instructions? PACK is followed by a NEWSTRUCT. When executing NEWSTRUCT, it is determined whether the operand is an integer or an array. If it is an array, it is converted into a structure.

File: src/neo-vm/OpCode.cs [hunk 1]
```diff
@@ -443,6 +443,10 @@ public enum OpCode : byte
         /// A map is taken from top of the main stack. The values of this map are put on top of the main stack.
         /// </summary>
         VALUES = 0xCD,
+        /// <summary>
+        /// A struct is removed from top of the main stack. Its elements are put on top of the main stack (in reverse order) and the struct size is also put on main stack.
```
Comment (igormcoelho) [inline src/neo-vm/OpCode.cs]: I guess this comment is reversed.... The purpuse is to extract elements from the main stack and put a struct, or am I mistaken?

File: tests/neo-vm.Tests/Tests/OpCodes/Arrays/PACKSTRUCT.json [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+                                "instructionPointer": 1,
+                                "nextInstruction": "RET"
+                            }
+                        ]
+                    }
+                }
+            ]
+        }
+    ]
+}
```
Comment (igormcoelho) [inline tests/neo-vm.Tests/Tests/OpCodes/Arrays/PACKSTRUCT.json]: Many hearts to you @shargon :heart:  this json testing idea is top best forever!

**Comment by igormcoelho** [comment]
I agree with @erikzhang , if `PACK`+`PACKSTRUCT` is used, the behavior will be the same. Syntax is consistent with current `PACKSTRUCT` (single input parameter, we just need to adapt the type, if not conflicting to any existing chain operation that used to fault), see operation 198: https://github.com/NeoResearch/yellowpaper/blob/master/sections/07_NeoVM_NeoContract.md
The only cost is having two opcodes, instead of one. If it helps compiling, another option is using a virtual opcode PACKSTRUCT, and then we can remove it on optimization process (transforming into two again).

File: tests/neo-vm.Tests/Tests/OpCodes/Arrays/PACKSTRUCT.json (hunk 2 above)
Comment (shargon) [reply tests/neo-vm.Tests/Tests/OpCodes/Arrays/PACKSTRUCT.json]: :) Thanks

**Comment by afragapane** [comment]
@erikzhang @igormcoelho @shargon @dicarlo2
The `PACT`+`NEWSTRUCT` solution should work for us.  I'll go ahead and get that implemented & tested in neo-one and then follow back up here.

Thanks for the feedback & suggestion!

**Comment by shargon** [comment]
I like the @erikzhang idea

**Pull request closed without merging**